<a href="https://colab.research.google.com/github/shreyash-99/EVA8/blob/main/Session%204%20Assignment%204%20of%20n.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Target 
1. Firstly implement stepLR to get consistent results
2. Reduce the size by a little bit to make it under 10k by changing some channel sizes
3. Will also try to increase to initial lr.

In [1]:
from __future__ import print_function
import torch
import torch.nn as nn
import torch.nn.functional as F
import torch.optim as optim 
from torchvision import datasets, transforms

In [2]:
train_transforms = transforms.Compose([transforms.RandomRotation((-7.0,7.0), fill = (1,)) ,  
                                      transforms.ToTensor(),
                                       transforms.Normalize((0.1307,), (0.3081,))
                                       ])
test_transforms = transforms.Compose([transforms.ToTensor(),
                                      transforms.Normalize((0.1307,), (0.3081, ))])


In [3]:
train = datasets.MNIST('./data', train = True, download = True, transform = train_transforms)
test = datasets.MNIST('./data', train = False, download = True, transform = test_transforms)

In [4]:
SEED = 1

cuda = torch.cuda.is_available()
print("CUDA available" , cuda)

#for Reproducibilty
torch.manual_seed(SEED)

if cuda:
  torch.cuda.manual_seed(SEED)

# setting the dataloader arguments for both test and train data
dataloader_args = dict(shuffle = True, batch_size = 128, num_workers = 4, pin_memory = True) if cuda else dict(shuffle = True, batch_size = 64)

train_loader = torch.utils.data.DataLoader(train , **dataloader_args)
test_loader = torch.utils.data.DataLoader(test, **dataloader_args)


CUDA available True


/usr/local/lib/python3.8/dist-packages/torch/utils/data/dataloader.py:554: UserWarning: This DataLoader will create 4 worker processes in total. Our suggested max number of worker in current system is 2, which is smaller than what this DataLoader is going to create. Please be aware that excessive worker creation might get DataLoader running slow or even freeze, lower the worker number to avoid potential slowness/freeze if necessary.
  warnings.warn(_create_warning_msg(


In [14]:
dropout_value = 0.07
class Net(nn.Module):
  def __init__(self):
    super(Net, self).__init__()
    ## INPUT BLOCK (BLOCK  1 )
    self.conv1 = nn.Sequential(
        nn.Conv2d(in_channels = 1, out_channels = 8, kernel_size = (3,3), bias = False),
        nn.BatchNorm2d(8),
        nn.ReLU(),
        nn.Dropout(dropout_value)   
    )# input - 28, output - 26
    self.conv2 = nn.Sequential(
        nn.Conv2d(in_channels = 8, out_channels = 16, kernel_size = (3,3), padding = 1, bias = False),
        nn.BatchNorm2d(16),
        nn.ReLU(),
        nn.Dropout(dropout_value)
    )#input - 26 , output - 26
    

    # TRANSITION BLOCK 
    self.conv3 = nn.Sequential(
        nn.Conv2d(in_channels = 16, out_channels = 8, kernel_size = (1,1), bias = False),
        # nn.BatchNorm2d(10),
        # nn.ReLU(),
        # nn.Dropout(dropout_value)
    )# input - 26 , output - 26
    self.pool1 = nn.MaxPool2d(2, 2)
    # input 26, output - 13

    # CONVOLUTION  BLOCK 2
    self.conv4 = nn.Sequential(
        nn.Conv2d(in_channels = 8, out_channels = 16 , kernel_size = (3,3), bias = False),
        nn.BatchNorm2d(16),
        nn.ReLU(),#input = 13 , output  = 11
        nn.Dropout(dropout_value)
    )
    self.conv5 = nn.Sequential(
        nn.Conv2d(in_channels = 16, out_channels = 16, kernel_size = (3,3), padding = 1, bias = False),
        nn.BatchNorm2d(16),
        nn.ReLU(),# input - 11 , output = 11
        nn.Dropout(dropout_value)
    )

    #   TRANSITION BLOCK 
    self.conv6 = nn.Sequential(
        nn.Conv2d(in_channels = 16, out_channels = 8, kernel_size = (1,1), bias = False),
        # nn.BatchNorm2d(10),
        # nn.ReLU(), # input - 11, output - 11
        # nn.Dropout(dropout_value)
    )
    # # self.pool2 = nn.MaxPool2d(2,2) #input 12, output 6

    # #   CONVOLUTION BLOCK 3

    self.conv7 = nn.Sequential(
        nn.Conv2d(in_channels = 8, out_channels = 8 , kernel_size = (3,3),bias = False),
        nn.BatchNorm2d(8),
        nn.ReLU(),#input = 11, output = 9
        nn.Dropout(dropout_value)
    )
    self.conv8 = nn.Sequential(
        nn.Conv2d(in_channels= 8, out_channels = 16, kernel_size = (3,3), bias = False),
        nn.BatchNorm2d(16),
        nn.ReLU(),#input = 9, output = 7
        nn.Dropout(dropout_value)
    )
    self.conv9 = nn.Sequential(
        nn.Conv2d(in_channels = 16, out_channels = 16, kernel_size = (3,3), bias = False),
        nn.BatchNorm2d(16),
        nn.ReLU(),
        nn.Dropout(dropout_value) 
    )#input - 7 , output - 5 


    #      CONVOLUTION BLOCK 4
    self.gap = nn.Sequential(
        nn.AvgPool2d(kernel_size = 5)
    )# output 1
    self.conv10 = nn.Sequential(
        nn.Conv2d(in_channels = 16, out_channels = 10, kernel_size = (1,1), bias = False)
    )
    


  def forward(self, x ):
    x = self.conv1(x)
    x = self.conv2(x)
    x = self.conv3(x)
    x = self.pool1(x)

    x = self.conv4(x)
    x = self.conv5(x)
    x = self.conv6(x)

    x = self.conv7(x)
    x = self.conv8(x)

    x = self.conv9(x)

    x = self.gap(x)

    x = self.conv10(x)


    x = x.view(-1,10)
    return F.log_softmax(x, dim = -1)


In [15]:
!pip install torchsummary
from torchsummary import summary
use_cuda = torch.cuda.is_available()
device = torch.device("cuda" if use_cuda else "cpu")
print(device)
model = Net().to(device)
summary(model, input_size = (1,28,28))

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
cuda
----------------------------------------------------------------
        Layer (type)               Output Shape         Param #
            Conv2d-1            [-1, 8, 26, 26]              72
       BatchNorm2d-2            [-1, 8, 26, 26]              16
              ReLU-3            [-1, 8, 26, 26]               0
           Dropout-4            [-1, 8, 26, 26]               0
            Conv2d-5           [-1, 16, 26, 26]           1,152
       BatchNorm2d-6           [-1, 16, 26, 26]              32
              ReLU-7           [-1, 16, 26, 26]               0
           Dropout-8           [-1, 16, 26, 26]               0
            Conv2d-9            [-1, 8, 26, 26]             128
        MaxPool2d-10            [-1, 8, 13, 13]               0
           Conv2d-11           [-1, 16, 11, 11]           1,152
      BatchNorm2d-12           [-1, 16, 11, 11]              32

In [16]:
from tqdm import tqdm

train_losses = []
test_losses = []
train_acc = []
test_acc = []

def train(model, device, train_loader, optimiser, epoch):
  model.train()
  pbar = tqdm(train_loader)
  correct = 0
  processed = 0
  for batch_idx, (data, target) in enumerate(pbar):
    #get samples
    data, target = data.to(device), target.to(device)

    #init
    optimiser.zero_grad()

    #prediction
    y_pred = model(data)

    #calculate loss
    loss = F.nll_loss(y_pred, target)
    train_losses.append(loss)

    #Backpropagation
    loss.backward()
    optimiser.step()

    # update pbar - tqdm

    pred = y_pred.argmax(dim = 1, keepdim = True) # gets the index of the max log-probabilirty
    correct += pred.eq(target.view_as(pred)).sum().item()
    processed += len(data)

    pbar.set_description(desc= f'Loss={loss.item()} Batch_id={batch_idx} Accuracy={100*correct/processed:0.2f}')
    train_acc.append(100*correct/processed)

def test(model, device, test_loader):
  model.eval()
  test_loss = 0
  correct = 0
  with torch.no_grad():
    for data, target in test_loader:
      data, target = data.to(device), target.to(device)
      output = model(data)
      test_loss += F.nll_loss(output, target, reduction='sum').item()  # sum up batch loss
      pred = output.argmax(dim=1, keepdim=True)  # get the index of the max log-probability
      correct += pred.eq(target.view_as(pred)).sum().item()

  test_loss /= len(test_loader.dataset)
  test_losses.append(test_loss)

  print('\nTest set: Average loss: {:.4f}, Accuracy: {}/{} ({:.2f}%)\n'.format(
    test_loss, correct, len(test_loader.dataset),
    100. * correct / len(test_loader.dataset)))
  
  test_acc.append(100. * correct / len(test_loader.dataset))


In [17]:
from torch.optim.lr_scheduler import StepLR

model =  Net().to(device)
optimizer = optim.SGD(model.parameters(), lr=0.03, momentum=0.9)
scheduler = StepLR(optimizer, step_size=6, gamma=0.1)


EPOCHS = 20
for epoch in range(EPOCHS):
    print("EPOCH:", epoch, "Learning Rate: ", scheduler.get_last_lr())
    train(model, device, train_loader, optimizer, epoch)
    scheduler.step()
    test(model, device, test_loader)

EPOCH: 0 Learning Rate:  [0.03]


Loss=0.13147200644016266 Batch_id=468 Accuracy=89.32: 100%|██████████| 469/469 [00:18<00:00, 25.69it/s]



Test set: Average loss: 0.0672, Accuracy: 9784/10000 (97.84%)

EPOCH: 1 Learning Rate:  [0.03]


Loss=0.026439359411597252 Batch_id=468 Accuracy=97.81: 100%|██████████| 469/469 [00:17<00:00, 26.10it/s]



Test set: Average loss: 0.0539, Accuracy: 9836/10000 (98.36%)

EPOCH: 2 Learning Rate:  [0.03]


Loss=0.10534457117319107 Batch_id=468 Accuracy=98.23: 100%|██████████| 469/469 [00:18<00:00, 25.39it/s]



Test set: Average loss: 0.0330, Accuracy: 9894/10000 (98.94%)

EPOCH: 3 Learning Rate:  [0.03]


Loss=0.03889813646674156 Batch_id=468 Accuracy=98.47: 100%|██████████| 469/469 [00:18<00:00, 25.96it/s]



Test set: Average loss: 0.0258, Accuracy: 9913/10000 (99.13%)

EPOCH: 4 Learning Rate:  [0.03]


Loss=0.18354378640651703 Batch_id=468 Accuracy=98.47: 100%|██████████| 469/469 [00:18<00:00, 25.46it/s]



Test set: Average loss: 0.0323, Accuracy: 9893/10000 (98.93%)

EPOCH: 5 Learning Rate:  [0.03]


Loss=0.023572811856865883 Batch_id=468 Accuracy=98.69: 100%|██████████| 469/469 [00:19<00:00, 24.57it/s]



Test set: Average loss: 0.0316, Accuracy: 9906/10000 (99.06%)

EPOCH: 6 Learning Rate:  [0.003]


Loss=0.017485691234469414 Batch_id=468 Accuracy=98.97: 100%|██████████| 469/469 [00:17<00:00, 26.34it/s]



Test set: Average loss: 0.0225, Accuracy: 9930/10000 (99.30%)

EPOCH: 7 Learning Rate:  [0.003]


Loss=0.013596788048744202 Batch_id=468 Accuracy=99.04: 100%|██████████| 469/469 [00:17<00:00, 26.21it/s]



Test set: Average loss: 0.0203, Accuracy: 9945/10000 (99.45%)

EPOCH: 8 Learning Rate:  [0.003]


Loss=0.07576064765453339 Batch_id=468 Accuracy=99.02: 100%|██████████| 469/469 [00:18<00:00, 25.99it/s]



Test set: Average loss: 0.0200, Accuracy: 9947/10000 (99.47%)

EPOCH: 9 Learning Rate:  [0.003]


Loss=0.09534329921007156 Batch_id=468 Accuracy=99.03: 100%|██████████| 469/469 [00:17<00:00, 26.43it/s]



Test set: Average loss: 0.0195, Accuracy: 9945/10000 (99.45%)

EPOCH: 10 Learning Rate:  [0.003]


Loss=0.0043024662882089615 Batch_id=468 Accuracy=99.07: 100%|██████████| 469/469 [00:17<00:00, 27.07it/s]



Test set: Average loss: 0.0183, Accuracy: 9950/10000 (99.50%)

EPOCH: 11 Learning Rate:  [0.003]


Loss=0.0062692477367818356 Batch_id=468 Accuracy=99.08: 100%|██████████| 469/469 [00:17<00:00, 26.54it/s]



Test set: Average loss: 0.0188, Accuracy: 9948/10000 (99.48%)

EPOCH: 12 Learning Rate:  [0.00030000000000000003]


Loss=0.01717514358460903 Batch_id=468 Accuracy=99.12: 100%|██████████| 469/469 [00:17<00:00, 26.85it/s]



Test set: Average loss: 0.0189, Accuracy: 9948/10000 (99.48%)

EPOCH: 13 Learning Rate:  [0.00030000000000000003]


Loss=0.002918594516813755 Batch_id=468 Accuracy=99.11: 100%|██████████| 469/469 [00:17<00:00, 27.02it/s]



Test set: Average loss: 0.0192, Accuracy: 9948/10000 (99.48%)

EPOCH: 14 Learning Rate:  [0.00030000000000000003]


Loss=0.006048992276191711 Batch_id=468 Accuracy=99.21: 100%|██████████| 469/469 [00:18<00:00, 25.06it/s]



Test set: Average loss: 0.0191, Accuracy: 9949/10000 (99.49%)

EPOCH: 15 Learning Rate:  [0.00030000000000000003]


Loss=0.0249591376632452 Batch_id=468 Accuracy=99.14: 100%|██████████| 469/469 [00:17<00:00, 26.47it/s]



Test set: Average loss: 0.0188, Accuracy: 9948/10000 (99.48%)

EPOCH: 16 Learning Rate:  [0.00030000000000000003]


Loss=0.033999521285295486 Batch_id=468 Accuracy=99.16: 100%|██████████| 469/469 [00:17<00:00, 26.42it/s]



Test set: Average loss: 0.0189, Accuracy: 9948/10000 (99.48%)

EPOCH: 17 Learning Rate:  [0.00030000000000000003]


Loss=0.03355110064148903 Batch_id=468 Accuracy=99.13: 100%|██████████| 469/469 [00:18<00:00, 25.98it/s]



Test set: Average loss: 0.0184, Accuracy: 9949/10000 (99.49%)

EPOCH: 18 Learning Rate:  [3.0000000000000004e-05]


Loss=0.11394136399030685 Batch_id=468 Accuracy=99.14: 100%|██████████| 469/469 [00:18<00:00, 25.81it/s]



Test set: Average loss: 0.0184, Accuracy: 9947/10000 (99.47%)

EPOCH: 19 Learning Rate:  [3.0000000000000004e-05]


Loss=0.0030106117483228445 Batch_id=468 Accuracy=99.11: 100%|██████████| 469/469 [00:17<00:00, 26.44it/s]



Test set: Average loss: 0.0182, Accuracy: 9947/10000 (99.47%)



Results:
1. Parameters: 9.3k
2. Best Test Accuracy: 99.50% (till 15th epoch)
3. Best Train Accuracy: 99.21% (till 15th epoch)

Conclusion:
1. Model is very good as it is not at all overfitting, results are good it achieves 99.4% accuracy from 7th epoch 
2. Dont know how reducing the but reducing the channel the size of channels(16->8) for first was better for the model.
3. Increasing the lr to 0.3 also helped to achieve the target fast and steplr helped it to reduced it so that the target doesnt overshoot.